[DDP Communication Hook] Update the typing info of comm hook output as well as some docstring

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62457
* #62456

Specify `Future[torch.Tensor]` as DDP communication hook return type, which should be explicitly a single tensor. The previous API takes a list that has a single tensor.

Note that now the typing info no longer accepts the internal type of `torch._C.Future`, which does not support torchscript and hence cannot support `Future[torch.Tensor]`.

Differential Revision: [D30007390](https://our.internmc.facebook.com/intern/diff/D30007390/)

Commit:
[DDP Communication Hook] Update the typing info of comm hook output as well as some docstring

Specify `Future[torch.Tensor]` as DDP communication hook return type, which should be explicitly a single tensor. The previous API takes a list that has a single tensor.

Note that now the typing info no longer accepts the internal type of `torch._C.Future`, which does not support torchscript and hence cannot support `Future[torch.Tensor]`.

Differential Revision: [D30007390](https://our.internmc.facebook.com/intern/diff/D30007390/)

[ghstack-poisoned]

Commit:
Update on "[DDP Communication Hook] Update the typing info of comm hook output as well as some docstring"


Specify `Future[torch.Tensor]` as DDP communication hook return type, which should be explicitly a single tensor. The previous API takes a list that has a single tensor.

Note that now the typing info no longer accepts the internal type of `torch._C.Future`, which does not support torchscript and hence cannot support `Future[torch.Tensor]`.

Differential Revision: [D30007390](https://our.internmc.facebook.com/intern/diff/D30007390/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
Thanks for the consolidation!

Commit:
Update on "[DDP Communication Hook] Update the typing info of comm hook output as well as some docstring"


Specify `Future[torch.Tensor]` as DDP communication hook return type, which should be explicitly a single tensor. The previous API takes a list that has a single tensor.

Note that now the typing info no longer accepts the internal type of `torch._C.Future`, which does not support torchscript and hence cannot support `Future[torch.Tensor]`.

Differential Revision: [D30007390](https://our.internmc.facebook.com/intern/diff/D30007390/)

[ghstack-poisoned]
